dbclient/MLFlowClient.py

```diff
@@ -198,3 +192,176 @@ def _cleanse_artifact_location(self, artifact_location):
+        checkpointer.write(run_id, new_run_id)
+        return new_run_id
+
+    def _create_run_and_log_helper(self, src_client, experiment_id, run_id, start_time, metrics, params, tags, ml_run_artifacts_dir, steps_checkpointer):
+        creation_checkpoint_key = run_id + "_create_run"
+        log_batch_checkpoint_key = run_id + "_log_batch"
+        run_artifacts_checkpoint_key = run_id + "_artifacts"
+
+        if not steps_checkpointer.contains(creation_checkpoint_key):
+            run = self.client.create_run(experiment_id, start_time=start_time, tags={})
+            steps_checkpointer.write(creation_checkpoint_key, run.info.run_id)
+
+        new_run_id = steps_checkpointer.get(creation_checkpoint_key)
+        metrics_obj = [Metric(key, val, start_time, step=0) for key, val in metrics.items()]
+        params_obj = [Param(key, val) for key, val in params.items()]
+
+        # We filter out the following tags for the imported runs.
+        DENY_LIST_TAGS = [
+            "mlflow.databricks.notebookID", "mlflow.databricks.notebookRevisionID", "mlflow.databricks.webappURL",
+            "mlflow.databricks.runURL", "mlflow.databricks.cluster.id", "mlflow.databricks.workspaceURL",
+            "mlflow.databricks.workspaceID", "mlflow.databricks.notebook.commandID", "mlflow.databricks.shellJobID",
+            "mlflow.databricks.shellJobRunID", "mlflow.databricks.jobID", "mlflow.databricks.jobRunID",
+            "mlflow.databricks.jobType", "mlflow.databricks.jobTypeInfo",
+            "mlflow.log-model.history", "mlflow.user", "mlflow.rootRunId"
+        ]
+        tags_obj = [RunTag(key, val) for key, val in tags.items() if key not in DENY_LIST_TAGS]
+
+        if not steps_checkpointer.contains(log_batch_checkpoint_key):
+            self.client.log_batch(new_run_id, metrics_obj, params_obj, tags_obj)
+            steps_checkpointer.write(log_batch_checkpoint_key, new_run_id)
+
+        if not steps_checkpointer.contains(run_artifacts_checkpoint_key):
+            self._download_and_upload_run_artifacts(src_client, run_id, new_run_id, ml_run_artifacts_dir)
+            steps_checkpointer.write(run_artifacts_checkpoint_key, new_run_id)
+
+        return new_run_id
+
+    def _download_and_upload_run_artifacts(self, src_client, old_run_id, new_run_id, ml_run_artifacts_dir):
+        # Download artifacts into temp directory: ml_run_artifacts/$run_id_temp and then upload to the destination
+        # workspace. After uploading, remove the temp dir along with the artifacts.
+        temp_artifact_dir = ml_run_artifacts_dir + old_run_id + "_temp/"
+        shutil.rmtree(temp_artifact_dir, ignore_errors=True)
+        os.makedirs(temp_artifact_dir)
+        artifacts = src_client.list_artifacts(old_run_id)
+        if len(artifacts) == 0:
+            return
+
+        logging.info(f"Downloading run artifacts for run_id: {old_run_id}")
+        src_client.download_artifacts(old_run_id, "", temp_artifact_dir)
```

Comment:
nit: may want to use a `try / finally` block here to make sure we always call `shutil.rmtree(temp_artifact_dir)` even if the upload attempt fails in the current thread

Reply:
good point. Done!